gazebosim/gz-sim
# Add system to an entity through Component Inspector

# 🎉 New feature

depends on #1524

## Summary

Adds a tool button to the Component Inspector for adding Systems to a model. When the button is clicked, a dialog pops up and lets you fill the `name`, `filename`, and `innerxml` fields of the System / plugin. Once confirmed, it invokes a service call introduced in  #1524 that adds the System to the selected model entity.

## Test it

1. launch the `diff_drive_no_plugin` test world:

    ```
    ign gazebo -v 4 ./src/ign-gazebo/test/worlds/diff_drive_no_plugin.sdf
    ```

1. Click on the vehicle model -  The Component Inspector should now show properties of the vehicle model

1. Click the `+` button in the toolbar in the Component Inspector - A dialog should pop out

1. Fill the `name` field with:

    ```
    ignition::gazebo::systems::DiffDrive
    ```


1. Fill the `filename` field with:

    ```
    ignition-gazebo-diff-drive-system
    ```

1. Fill the `innerxml` field with:

    ```
        <left_joint>left_wheel_joint</left_joint>
        <right_joint>right_wheel_joint</right_joint>
        <wheel_separation>1.25</wheel_separation>
        <wheel_radius>0.3</wheel_radius>
        <max_linear_acceleration>1</max_linear_acceleration>
        <min_linear_acceleration>-1</min_linear_acceleration>
        <max_angular_acceleration>2</max_angular_acceleration>
        <min_angular_acceleration>-2</min_angular_acceleration>
        <max_linear_velocity>0.5</max_linear_velocity>
        <min_linear_velocity>-0.5</min_linear_velocity>
        <max_angular_velocity>1</max_angular_velocity>
        <min_angular_velocity>-1</min_angular_velocity>    
    ```


1.  Click Ok to add the System to the model - You should see the `+ System Plugin Info` appear in the Component Inspector to indicate that the model now has a System.

1.  Hit Play - The plugin should be running and read to accept vel commands

1. Publish a message to move the vehicle:

    ```
    ign topic -t "/model/vehicle/cmd_vel" -m ignition.msgs.Twist -p "linear: {x: 0.5}, angular: {z: 0.05}"
    ```

![add_system_gui](https://user-images.githubusercontent.com/4000684/174926843-b9ce5ede-06c9-4aad-8a12-f1c391953a77.gif)



## Checklist


**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.
